### DISCUSSION TITLE
Collaboration/contribution

### DISCUSSION THREAD

**Discussion by tomflint22** [chaos-polymtl/lethe#1058] [Ideas]
Hi,

I just stubbled across this project. I've just installed and planning on playing around over the next few weeks. At first look it looks like you have some laser heating models in your VOF implementation.

I recently published an openfoam solver with ray-tracing for laser-substrate interactions(laserbeamfoam):

https://www.sciencedirect.com/science/article/pii/S2352711023003084

https://github.com/micmog/LaserbeamFoam

The laser ray-tracing algorithm is really general (laserHS.H in the repo). I think this could be really easily incorporated into Lethe, and I'd be very happy to help with this if someone from your team was willing to hold my hand through the solver creation/modification process. As I say I only just installed (running the tests now and all is looking good). Maybe after I've played around, and have a feel for how everything works we could exchange a few messages/pull requests. I'd love to learn how to implement my frameworks in your project.

In the longer term, our group at The University of Manchester develop various continuum mechanics frameworks (multi-hase field solvers for multi-component solidification, heterogeneous MHD solvers, Eulerian plasticity and crystal plasticity) that I would be super interested in incorporating into Lethe. I've done everything on-top of openfoam until now but looking to diversify from there.

Anyway, just an initial thought on the ray-tracing, but I'd love to get to know your code and be able to contribute.

All the best,
Tom

**Comment by hepap** [comment]
Hi Tom,

Thanks for reaching out! We do have continuous laser models in Lethe (you can find a brief description here: https://lethe-cfd.github.io/lethe/documentation/parameters/cfd/laser_heat_source.html).

The ray tracing algorithm is one of our mid-term goals, and it would be really nice to talk about it when you are all set up with Lethe. In the meantime, do not hesitate to contact us if you have any questions/comments about Lethe. We will also be looking at your ray tracing algorithm to see how it might be integrated.

One of our goals with Lethe is to have a fully verified and validated LPBF simulation tool at the melt pool scale. We are currently developing benchmark 1D, 2D, and 3D cases to perform code-to-code comparison and validation to access the accuracy and adequacy of different solvers. If this is something you would be interested in, I can provide the input parameters!

Thanks again for reaching out!

We look forward to speaking with you,
Hélène

**Reply by tomflint22** [reply]
Ace, well our code is validated against the Sen and Davis case for the Marangoni implementation, the Gallium melting case (that I see you already have) for the melting and latent heat stuff, the Tan case (Gallium case with a free surface) again for melting and well VOF stuff I guess. We've also got a lot of data for EB welds that our solvers perform really well against. Also, we have a higher fidelity version of the code that will be part of the repo eventually, but isn't released yet. You can find some papers on hat implementation here where wee explicitly consider the volumetric dilation between the liquid/vapour pairs in multi-component substrates:

https://doi.org/10.1016/j.ijheatmasstransfer.2023.124262

https://doi.org/10.1016/j.ijheatmasstransfer.2022.122985

To be honest, the thing that would interest me the most is directly contributing to get more things implemented in your code (e.g. the ray tracing algorithm as I mentioned), rather than just benchmarking. We also have some novel MHD frameworks:
https://www.nature.com/articles/s41598-021-97177-8

and Eulerian solid dynamics descriptions that I would be interested in sharing/implementing in lethe for my own interest and to diversify away from OpenFoam. If this sounds like something that would be interesting to your group, I think I can bring something that may be useful. As I say, I would just need some time from your team to be able to ask questions. I've implemented things in dealii before (well prisms -pf that is built on dealii like lethe): 

https://github.com/prisms-center/phaseField/tree/additive_manufacturing/applications/_additive_manufacturing

We'll be making our Openfoam poly-crystal (Eulerian) plasticity implementations, and multi-component multi-phase field implementations (dendritic solidification with flow/solute partitioning etc) open-source within the next 6-12 months so you'll see more there and if that work is of interest to be included in lethe for you. By the way, when I say we I mean myself, Pratheek Shanthraj and Philip Cardiff; that's the small team we have that does all this stuff.

I don't want to be annoying. Basically I'm really keen to contribute my stuff into lethe over the medium to long term, and just wondering if this is of interest. And if it is if you'd like a zoom call at some point to discuss.

I'm sure you've got all the expertise within your group, so maybe there's no appetite, and that's totally fine. It's just nice to come across such a nice project. It's how I felt when I first came across OpenFoam and stopped working on my own finite difference codes and started doing everything in that.

All the best,

Tom

https://github.com/lethe-cfd/lethe/assets/17575203/e45ff94b-d186-4723-aeb3-27e61067da49

**Reply by tomflint22** [reply]
P.S. I ran a few lethe simulations without any issues. The install documentation is first rate!

I'll be finding time next week to start digging into how everything works, but is the idea behind the structure that the solver, lethe-fluid, is called and based on flags in the .prm file other physics modules are also activated? Like the energy conservation equation or cahn Hilliard? 

Obviously I just ran lethe-fluid on the cahn-hilliard bubble case and it called that physics solver. I guess what I'd like to do is add these physics solvers if that makes sense.

Anyway, enjoy your weekends. Looking forward to hearing back, and hearing any thoughts on collaboration.

All the best,
Tom

**Reply by blaisb** [reply]
Dear @tomflint22,
Thank you for your interested. It will be our pleasure to help you integrate whatever you wish to integrate within the software, we have a very relaxed policy on adding user contributions :).

The way it works is relatively straightforward. Lethe is made of three main categories of solvers
Fluid:
lethe-fluid
lethe-fluid-block
lethe-fluid-matrix-free

Particles:
lethe-particles

FEM-DEM:
lethe-fluid-particles

The lethe-fluid is the main CFD solver, with lethe-fluid-matrix-free our new matrix-free implementation that scales very well and lethe-fluid-block kinda being a pet project of mine with which I have never managed to be satisfied with (Block preconditioners for stabilized equations are hard to establish / it is not my main area of expertise).

The multiphysics features of Lethe are just enabled by setting up the parameter files adequately. We have multiple parameters and post-processing capabilities which are all described in the documentation (https://lethe-cfd.github.io/lethe/documentation/parameters/parameters.html). The models are in constant evolution.

Adding a new physics requires some work, but there is a nice API which is called the multiphysics interface which enables different physics to exchange and trade data.

We welcome all contributions. This was started around 4 years ago as a way to make FEM-based CFD through deal.II more accessible to everyone. The main focus is on keeping it scalable, but also keeping everything open and well-documented so that everyone can use what we do. Feel free to reach out if you need more help, I'm always glad to help people get onboard. The goal of this is for it to remain fun :)

Take care
BB

**Reply by tomflint22** [reply]
Hi Bruno,

thanks very much for taking the time to reply. It's a really cool project you've got going here!

Thanks for the explanations. I really appreciate it. It's nice to hear your philosophy of keeping things fun! That, along with a commitment to open-science and open-source tools, is our attitude too. Hopefully this could develop into a nice collaboration. I have students starting in September and if I manage to make good progress, they could be additional people who could help in the porting of the various features and tools we have into your code.

Going forward, what is the best way to ask questions? Should I start threads on this discussion board page?

In any-case, I'll start playing around this week, probably breaking most things. I'll try and look through the documentation as best I can before asking questions, but apologies in advance if I ask too many basic questions......

All the very best,

Tom

**Reply by blaisb** [reply]
Dear Tom,
Asking questions here would be the best way to proceed. We are trying to support the discussion part of our Github website more adequately. Initially we thought about starting a forum, but then realized that this meant another website to support (on top of the doxygen, the documentation, the repository, the CI, etc.), so we decided we would keep on using the discussion pane in Github for this type of activity.

Feel free to ask as many questions as you want, we will be glad to answer them.

As for other ways to collaborate, feel free to keep us informed of our activities. There are often chances for internships (with scholarships available from Canadian funding agencies) and I am always very happy to welcome interns from other universities that want to work with/on Lethe.

Good luck  in your endeavours! It's normal if everything breaks, this is still a work in progress :).
